lightclient/src/platform/wasm_socket.rs

```diff
@@ -134,9 +134,9 @@ impl WasmSocket {
         });
         socket.set_onerror(Some(error_callback.as_ref().unchecked_ref()));
 
-        let close_callback = Closure::<dyn FnMut(_)>::new({
+        let close_callback = Closure::once_into_js({
```

Comment:
The docs for `once_into_js` say:

```
Convert a FnOnce(A...) -> R into a JavaScript Function object.

If the JavaScript function is invoked more than once, it will throw an exception.

Unlike Closure::once, this does not return a Closure that can be dropped before the function is invoked to deallocate the closure. The only way the FnOnce is deallocated is by calling the JavaScript function. If the JavaScript function is never called then the FnOnce and everything it closes over will leak.
```

So, since some of these callbacks might never fire, I think we'd be gradually leaking them here for any connection we make that doesn't fire one of the callbacks.

I wonder if the errors we got happened because previously, if `Self` is dropped before one of the JS websocket callbacks is fired, then it would have decallocated the closures and we'd get an error in JS land instead.

But anyway, from reading [the spec](https://websockets.spec.whatwg.org/#interface-definition), it sounds like we:
- May never fire `onopen` if connection can't be opened (but will fire it max once)
- May never fire `onerror` or `onmessage` if no errors/messages sent (not certain if `onerror` will only fire once or not)
- Will always eventualy fire `onclose` (max once).

So if I read right, I wonder whether we can do some cleanup in `onclose` closure to ensure that all of the other closures are dropped. Then, we can otherwise just leak them all so that nothing can lead them to being dropped otherwise.

Reply:
That makes sense!

I've found another way, which I believe to be more straight forward than performing clean-up on the `onclose`:
- keep the `Closure::new()` around for each callback
- while dropping the object:
  - close the socket if in: OPEN or CONNECTING states
  - set closures on the web_sys::WebSocket as None

This shall clean-up all memory associated with the rust closures, and deny the javascript real access to the dropped objects